dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/NativeRelationAnalysis.java

```diff
@@ -877,6 +877,7 @@ public MutableKnowledge visitReadFrom(ReadFrom rf) {
                 for (Event r : g.getRange()) {
                     may.removeIf((e1, e2) -> e2 == r);
                 }
+                may.addAll(g);
```

Comment:
I would change this code to do `may.removeIf((e1, e2) -> e2 == r && !g.contains(e1, e2))`.
The reason is that this way, if we actually do add an rf-edge between events that may not read from each other, we get the desired mutual exclusion.
Your solution, in some sense, introduces back imprecision in such cases (must-not may suddently become may).